File: csharp/ISBN13/ISBN13.cs
using System.Numerics;
using System.Globalization;
namespace ISBN13;

public class ISBN13
{
    public ISBN13(string isbn)
    {
        this.ISBN = isbn;
    }

    public string ISBN { get; }

    public bool IsValid()
    {
        if (this.ISBN.Length < 13)
        {
            return false;
        }
        if (this.ISBN.Length > 13)
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[0]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[1]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[2]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[3]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[4]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[5]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[6]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[7]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[8]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[9]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[10]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[11]))
        {
            return false;
        }
        if (!new List<char>() { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' }.Contains(this.ISBN[12]))
        {
            return false;
        }
        if (((this.ISBN[0] - (int)'0')
            + 3 * (this.ISBN[1] - (int)'0')
            + (this.ISBN[2] - (int)'0')
            + 3 * (this.ISBN[3] - (int)'0')
            + (this.ISBN[4] - (int)'0')
            + 3 * (this.ISBN[5] - (int)'0')
            + (this.ISBN[6] - (int)'0')
            + 3 * (this.ISBN[7] - (int)'0')
            + (this.ISBN[8] - (int)'0')
            + 3 * (this.ISBN[9] - (int)'0')
            + (this.ISBN[10] - (int)'0')
            + 3 * (this.ISBN[11] - (int)'0')
            + (this.ISBN[12] - (int)'0')
        ) % 10 == 0)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
}

File: python/isbn13/isbn13.py
class ISBN13:
    def __init__(self, isbn):
        self.isbn = isbn

    def is_valid(self) -> bool:
        # see https://en.wikipedia.org/wiki/ISBN#ISBN-13_check_digit_calculation
        if len(self.isbn) < 13:
            return False
        if len(self.isbn) > 13:
            return False
        if self.isbn[0] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[1] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[2] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[3] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[4] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[5] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[6] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[7] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[8] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[9] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[10] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[11] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if self.isbn[12] not in ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]:
            return False
        if (
            int(self.isbn[0])
            + 3 * int(self.isbn[1])
            + int(self.isbn[2])
            + 3 * int(self.isbn[3])
            + int(self.isbn[4])
            + 3 * int(self.isbn[5])
            + int(self.isbn[6])
            + 3 * int(self.isbn[7])
            + int(self.isbn[8])
            + 3 * int(self.isbn[9])
            + int(self.isbn[10])
            + 3 * int(self.isbn[11])
            + int(self.isbn[12])
        ) % 10 == 0:
            return True
        else:
            return False
